### ISSUE TITLE
 Combine Resolve, ResolveAll, and Invoke into a Locate func

### ISSUE THREAD

**Issue by breerly** [uber-go/dig#34]
When you ask the container to resolve dependencies, that is called the "service locator" pattern.

Combining `Resolve`, `ResolveAll`, and `Invoke` into a `Locate` func will streamline the experience:

```go
// locate 1 or more types
var logger *zap.Logger
var dispatcher *yarpc.Dispatcher
container.Locate(&logger, &dispatcher)

// same as above, using a func
container.Locate(func(logger *zap.Logger, dispatcher *yarpc.Dispatcher) {
 
})
```

This has the benefit of:

* Being more specific, giving a nod to the pattern actually being used here
* Reduces API surface and simplifies using DIG

**Comment by breerly** [comment]
(p.s. this issue is to *discuss* - had the idea this A.M. and wanted to jam on it)

**Comment by breerly** [comment]
In fact, I'm even in favor of *only* having this form:

```go
container.Locate(func(logger *zap.Logger, dispatcher *yarpc.Dispatcher) {
 
})
```
I don't really see the benefit to the original `Resolve` form with the above.

**Comment by glibsm** [comment]
This is definitely something I've thought about once we solidified the api for `Invoke`. @anuptalwalkar and I chatted offline and realized we can completely remove all forms of `Provide` and replace with this one thing.

I will let him chime in, since he had some thoughts and why we shouldn't do it.

**Comment by anuptalwalkar** [comment]
I like having three APIs - `Provide`, `Resolve` and `Invoke` in the form of variadic arguments since each have its own use. We should remove all Must functions, and All functions.

I like the naming of `Invoke` rather than locate, simply because `Invoke` comes with a sense of providing a function (be it anonymous func), `Locate` sounds more like an API on an object. 

My main issue with only relying on Invoke was all the locking around the functionality, which is not the case anymore with the refactor. That being said, we should keep both `Resolve` and `Invoke`. The difference between `Invoke` and `Resolve` is the need to have everything wrapped in a function. What if I just need logger from the graph? Following is not very intuitive- 
```go
var l *zap.logger 
container.Invoke(func(logger *zap.Logger) {
   l = logger
})
```

`Provide` is allowing us to add multiple objects, constructors to the graph. Again, wrapping into anonymous function or invoking a function is not very intuitive.

**Comment by breerly** [comment]
@anuptalwalkar, ok I buy that the first form is also useful, this is pretty ugly:

```go
var l *zap.logger 
container.Invoke(func(logger *zap.Logger) {
   l = logger
})
```

Can you compromise with making both forms available over `Locate`?

```
var logger *zap.Logger
container.Locate(logger)

container.Locate(func(logger *zap.Logger) {

})
```

I hear you on `Invoke` making it obvious that the func will be executed, but I really do think that this naming of `Locate` will make lots of sense:

> Container.Locate() locates types out of the container, it takes a pointer to the type or a func which will be invoked with provided types as parameters

**Comment by breerly** [comment]
In other words, I *really* want to see only 2 functions on the public API: `Provide` and `Locate`.

**Comment by glibsm** [comment]
I'm trying not to drag this into the naming conversation too much, but `Locate` to me sounds like a map lookup by key, rather than a more sophisticated process of recursively instantiating dependencies and all that stuff. 

I do understand that there is an actual pattern called the locator pattern, but we're at this point losing all connection to the dependency injection graphs.

**Comment by breerly** [comment]
Ok, how about calling it `Resolve`?

```go
var logger *zap.Logger
container.Resolve(logger)

container.Resolve(func(logger *zap.Logger) {

})
```

**Comment by glibsm** [comment]
My main point is about multiplexing based on the input parameters. I think it's very confusing from the perspective of documentation and new users coming in to see a function that takes in a variadic array of `interface{}` that can multiplex it's behaviour based on the type of the parameter being passed in.

Do we have other places in the go ecosystem that do different things based on the type of passed in parameters?

**Comment by breerly** [comment]
@glibsm in this case I'd like to see us show how to use dig using a really succinct README.

I think for this problem area, less functions that do more things is actually the right balance to strike.

**Comment by breerly** [comment]
ok @anuptalwalkar convinced me: let's start by removing the `*All` funcs so that we are left with:

* `Provide`
* `Resolve`
* `Invoke`

**Comment by breerly** [comment]
Closed in favor of #40
